### PR TITLE
JIP-26 Voting

### PR DESCRIPTION
**Pull request by dryu99**
I can't feel my arm AAAAAA

## Changes
- Allow users to vote for posts
  - Implemented necessary(?) endpoints
  - Didn't change anything in the schemas. 
  - When a user logs in (via cookie or dialog), we just query for user voted posts which we use on the client to build a data structure that helps us update upvote/downvote UI quickly.
- Display auth errors if users try to vote for posts without being logged in
- Fixed some jank with the way we were fetching complete post data in individual post view 

@srhsrhsrhsrh when fetching user voted posts in the dialog view you can probably just reuse/spiceup `getVotedPostsByUserID` niceu

### CODE REVIEW

File: packages/client/src/redux/slices/post-slice.ts [hunk 1]
```diff
@@ -187,6 +172,21 @@ export const postSlice = createSlice({
...
+      );
+
+      if (postToUpdate) {
+        state.posts = state.posts.map((post) =>
+          post._id === action.payload._id ? action.payload : post,
+        );
+      } else {
+        // this case solely exists to handle workflow where user navs to single post view via url
+        state.posts = [action.payload];
+        state.currPostIndex = 0;
```
Comment (srhsrhsrhsrh) [inline packages/client/src/redux/slices/post-slice.ts:187]: nice

File: packages/client/src/redux/slices/user-slice.ts [hunk 2]
```diff
@@ -12,9 +12,18 @@ export type UserState = {
   username?: string;
   email?: string;
   reputation?: number;
-  // password?: string (shouldn't be stored on client)
   posts: Post[];
-  votedPosts: { [id: string]: { upvoted: boolean } }; // TODO idk if this is best way to do it
+  votedPosts: { [id: string]: { upvoted: boolean } };
```
Comment (srhsrhsrhsrh) [inline packages/client/src/redux/slices/user-slice.ts:16]: he decided it was the best way to do it

**Review by srhsrhsrhsrh — APPROVED** [review]

File: packages/server/src/data/db/db-operations/post-ops.js [hunk 3]
```diff
@@ -102,7 +104,77 @@ const getPostsByUserID = async (userId, sortType) => {
...
+  return updatedPost
+    ? { ...updatedPost.toJSON(), comments: undefined } // TODO use $project instead
+    : undefined;
+};
+
+const downvotePost = async (postId, userId) => {
+  const updateSpec = {
+    $push: { downvoters: userId },
+    $pull: { upvoters: userId },
+    $inc: { numDownvotes: 1 },
```
Comment (jpfranci) [inline packages/server/src/data/db/db-operations/post-ops.js:163]: Wait what if they already downvoted

File: packages/server/src/data/db/db-operations/post-ops.js [hunk 4]
```diff
@@ -102,7 +104,77 @@ const getPostsByUserID = async (userId, sortType) => {
...
+  };
+
+  const post = await Post.findById(postId);
+  if (post.upvoters.includes(userId)) {
+    updateSpec.$inc.numUpvotes = -1;
+  }
+
+  const updatedPost = await Post.findByIdAndUpdate(postId, updateSpec, {
```
Comment (jpfranci) [inline packages/server/src/data/db/db-operations/post-ops.js:171]: I think we might have to change the where clause to use `$nin` post in downvoters. 

https://docs.mongodb.com/manual/reference/operator/query/nin/

File: packages/client/src/redux/slices/user-slice.ts (hunk 2 above)
Comment (jpfranci) [reply packages/client/src/redux/slices/user-slice.ts:16]: This doesn't work for me when I first load the page. Can't we encode isUpvoted/isDownvoted into every post as it's retrieved?

File: packages/server/src/data/db/db-operations/post-ops.js [hunk 5]
```diff
@@ -102,7 +104,77 @@ const getPostsByUserID = async (userId, sortType) => {
...
+    $project: {
+      title: true,
+      body: true,
+      date: true,
+      numUpvotes: true,
+      numDownvotes: true,
+      username: true,
+      tag: true,
+      score: { $subtract: ["$numUpvotes", "$numDownvotes"] },
+      isMature: true,
```
Comment (jpfranci) [inline packages/server/src/data/db/db-operations/post-ops.js:130]: Basically can just add this:

isUpvoted: { $in: [userId, "$upvoters"] },
isDownvoted: { $in: [userId, "$downvoters"] },

File: packages/server/src/data/db/db-operations/post-ops.js (hunk 4 above)
Comment (jpfranci) [reply packages/server/src/data/db/db-operations/post-ops.js:171]: Also we should just return the same fields as above for projection, don't need everything:
https://stackoverflow.com/questions/27302398/mongoose-how-to-project-protected-fields-on-update

**Comment by jpfranci** [comment]
Opened up this PR with changes that I was talking about. 

https://github.com/jpfranci/Alpacow/pull/31

File: packages/client/src/redux/slices/user-slice.ts (hunk 2 above)
Comment (dryu99) [reply packages/client/src/redux/slices/user-slice.ts:16]: > he decided it was the best way to do it

narrator: it was not the best way to do it

**Pull request merged**